File: copy_big_file.py
import time
from pathlib import Path

DEST_FOLDER = Path("E:\\tmp")
INPUT = DEST_FOLDER / "big.bin"
OUTPUT = DEST_FOLDER / "big_out.bin"
CHUNK_BYTES = 64 * 1024**2  # 64 MiB


def main() -> None:
    buf = bytearray(CHUNK_BYTES)
    mv = memoryview(buf)
    read_time = 0.0
    write_time = 0.0
    total = 0

    start = time.monotonic()
    with open(INPUT, "rb") as f_in, open(OUTPUT, "wb") as f_out:
        while True:
            t = time.monotonic()
            n = f_in.readinto(buf)
            read_time += time.monotonic() - t
            if n == 0:
                break
            t = time.monotonic()
            f_out.write(mv[:n])
            write_time += time.monotonic() - t
            total += n

    elapsed = time.monotonic() - start
    total_mb = total / (1024 * 1024)
    print(f"Copied {total:,} bytes in {elapsed:.2f}s")
    print(f"  read:  {read_time:.2f}s ({total_mb / read_time:.1f} MB/s)")
    print(f"  write: {write_time:.2f}s ({total_mb / write_time:.1f} MB/s)")


if __name__ == "__main__":
    main()

File: copy_big_file.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <time.h>

#define INPUT       "E:\\tmp\\big.bin"
#define OUTPUT      "E:\\tmp\\big_out.bin"
#define CHUNK_BYTES (64ULL * 1024 * 1024)   /* 64 MiB */

int main(void) {
    uint8_t *buf = (uint8_t *)malloc(CHUNK_BYTES);
    if (!buf) {
        fprintf(stderr, "malloc failed\n");
        return 1;
    }

    FILE *f_in = fopen(INPUT, "rb");
    if (!f_in) {
        fprintf(stderr, "fopen input failed: %s\n", INPUT);
        free(buf);
        return 1;
    }
    FILE *f_out = fopen(OUTPUT, "wb");
    if (!f_out) {
        fprintf(stderr, "fopen output failed: %s\n", OUTPUT);
        fclose(f_in);
        free(buf);
        return 1;
    }

    double read_time = 0.0;
    double write_time = 0.0;
    unsigned long long total = 0;

    clock_t start_total = clock();
    while (1) {
        clock_t t = clock();
        size_t n = fread(buf, 1, CHUNK_BYTES, f_in);
        read_time += (double)(clock() - t) / CLOCKS_PER_SEC;
        if (n == 0) {
            break;
        }

        t = clock();
        fwrite(buf, 1, n, f_out);
        write_time += (double)(clock() - t) / CLOCKS_PER_SEC;

        total += n;
    }
    double elapsed = (double)(clock() - start_total) / CLOCKS_PER_SEC;

    fclose(f_in);
    fclose(f_out);
    free(buf);

    double total_mb = (double)total / (1024.0 * 1024.0);
    printf("Copied %llu bytes in %.2fs\n", total, elapsed);
    printf("  read:  %.2fs (%.1f MB/s)\n", read_time, total_mb / read_time);
    printf("  write: %.2fs (%.1f MB/s)\n", write_time, total_mb / write_time);
    return 0;
}
